Model/DutyCalculator.php

```diff
@@ -116,84 +116,333 @@ public function __construct(
+        //should not that be based on corresponding admin setting?
+        $this->checkoutSession->setReachDuty($duty_adjusted);
+
+        if ($apply || !$this->getIsOptional($address->getCountryId())) {
+            //checkbox selection was duty should be applied
+            //or applying duty is a must for that country
+            $baseCurrency = $this->storeManager->getStore()->getBaseCurrency();
+            $rate = $baseCurrency->getRate($baseCurrency->getCode());
+            $baseDuty = $duty_adjusted / $rate;
+            $quote->setBaseReachDuty($baseDuty);
+            $this->checkoutSession->setBaseReachDuty($baseDuty);
+            $quote->setReachDuty($duty_adjusted);
+            $this->checkoutSession->setReachDuty($duty_adjusted);
+            $quote->setDhlQuoteId($response['quoteId']);
+            $quote->setDhlBreakdown(json_encode($response['feeTotals']));
+            $this->checkoutSession->setDhlBreakdown($quote->getDhlBreakdown());
+            $this->checkoutSession->setApply(true); //checkbox selection /corresponding passed value
+            //implies that duty should be applied
+            $this->_logger->debug('Apply block immediately after DHL call');
+        } else {
+            //checkbox selection or parameter passed indicates that duty should not be applied
+            $quote->setBaseReachDuty(0);
+            $quote->setReachDuty(0);
+            $this->checkoutSession->setBaseReachDuty(0);
+            $this->checkoutSession->setReachDuty($duty_adjusted);
+            $quote->setDhlQuoteId('');
+            $this->checkoutSession->setDhlQuoteId('');
+            $quote->setDhlBreakdown('');
+            $this->checkoutSession->setDhlBreakdown('');
+            $this->checkoutSession->setApply(false);
+            $this->_logger->debug('Do not Apply block immediately after DHL call');
+        }
+
+        $quote->save();
+        $this->response->setSuccess(true);
+        $this->response->setDuty($duty_adjusted);
+        $this->response->setIsOptional($this->getIsOptional($address->getCountryId()));
+    }
+
+    /**
+     * Deals with a case when DHL API call does not return a value for duty
+     * @param Magento\Framework\App\Response $response
+     */
+    public function fillOutQuoteAndSessionOnError($response)
+    {
+        $this->response->setSuccess(false);
+        $this->response->setDuty(0);
+
+        if (isset($response['message'])) { //error message from DHL duty and tax api call?
+            //if yes then do we want to call the API again?
+            //Assuming we do; we are resetting previous country and region value so that we can reenter the blcok to make
+            //DHL API call
+            $this->checkoutSession->setResponseErrorMessage($response['message']);
+            $this->response->setErrorMessage($response['message']);
+            $this->checkoutSession->setPrevCountry('');
+            $this->checkoutSession->setPrevRegion('');
```

Comment:
I see that only the session is being reset, should the quote be reset and saved as well?

Reply:
Excellent point @lizzynibali . (As per the logic/flow) for country under such a situation (duty and shipment not allowed as per some admin config)  control flow is returning back immediately and no decision (unless I missed something due to still learning the system) is made there based on what is in saved  in quote (even if it is stale and has value specific to a prior country selection where duty is relevant). That said I do not know whether there is any report generation happening somewhere else in the system based on saved state (of quote) in the database or not. In that case I should cleanup saved state too. Thanks.

Reply:
If possible I would prefer the function name to be changed to be accurate (for ex. **fillOutResponseAndSessionOnError**) since nothing is changing with the Quote. 

Re-reading the function I don't think we need to set the quote since the response is coming back with an error, but if we do then we'll run into that issue when we get there. Any suggestions for resetting the quote would just be guessing and I don't want to recommend code changes off of a guess.

Let me know if that makes sense to you too

Reply:
Changed the name as part of commit b9d08fa
Thanks.

Reply:
thank you!